contracts/GenericSwap.sol

```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import { TokenCollector } from "./abstracts/TokenCollector.sol";
+import { EIP712 } from "./abstracts/EIP712.sol";
+import { IGenericSwap } from "./interfaces/IGenericSwap.sol";
+import { IStrategy } from "./interfaces/IStrategy.sol";
+import { GeneralAsset } from "./libraries/GeneralAsset.sol";
+import { SignatureValidator } from "./libraries/SignatureValidator.sol";
+
+contract GenericSwap is IGenericSwap, TokenCollector, EIP712 {
+    using SafeERC20 for IERC20;
+    using GeneralAsset for address;
+
+    bytes32 public constant GS_DATA_TYPEHASH = 0x9a6d9096b182513baa520ad9d0766c6e70d0637fcf40521146c04435396a0fdf;
+
+    mapping(bytes32 => bool) private filledSwap;
+
+    /*
+        keccak256(
+            abi.encodePacked(
+                "GenericSwapData(",
+                "address inputToken,",
+                "address outputToken,",
+                "uint256 inputAmount,",
+                "uint256 minOutputAmount,",
+                "address receiver,",
+                "uint256 deadline,",
+                "bytes inputData,",
+                "bytes strategyData,",
+                "uint256 salt",
+                ")"
+            )
+        );
+        */
+
+    constructor(address _uniswapPermit2) TokenCollector(_uniswapPermit2) {}
+
+    /// @param strategy The strategy contract
+    /// @param swapData Swap data
+    /// @return returnAmount Output amount of the swap
+    function executeSwap(IStrategy strategy, GenericSwapData calldata swapData) external payable override returns (uint256 returnAmount) {
+        return _executeSwap(strategy, swapData, msg.sender);
+    }
+
+    /// @param strategy The strategy contract
+    /// @param swapData Swap data
+    /// @return returnAmount Output amount of the swap
+    function executeSwap(
+        IStrategy strategy,
+        GenericSwapData calldata swapData,
+        address taker,
+        uint256 salt,
+        bytes calldata takerSig
+    ) external payable override returns (uint256 returnAmount) {
+        bytes32 swapHash = getEIP712Hash(_getGSDataHash(swapData, salt));
+        require(!filledSwap[swapHash], "already filled");
+        if (!SignatureValidator.isValidSignature(taker, swapHash, takerSig)) revert InvalidSignature();
+        filledSwap[swapHash] = true;
+        return _executeSwap(strategy, swapData, taker);
+    }
+
+    function _executeSwap(
+        IStrategy strategy,
+        GenericSwapData calldata swapData,
+        address taker
```

Comment:
Q: This role is more like a `maker` (or `owner`), right? Since we collect the swapping token from him, he is the one who makes / owns the order, and the strategy is more like a taker to fulfill the order for him.

Reply:
@doublespending You can check if you have any other inputs

Reply:
Since the quote is provided by AMM protocols (or RFQ/LimitOrder), I still think it's considered as `taker` here. The taker is the one that accept the quote, which is given by maker.